src/lib/cache_lib/impl.ml

```diff
@@ -0,0 +1,236 @@
+open Async_kernel
+open Core_kernel
+open Or_error.Let_syntax
+
+module type Inputs_intf = sig
+  val handle_unconsumed_cache_item :
+    logger:Logger.t -> cache_name:string -> unit
+end
+
+module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
+  module rec Cache : sig
+    include Intf.Cache.S with module Cached := Cached
+
+    val remove : 'elt t -> 'elt -> unit Or_error.t
+  end = struct
+    type 'a t = {name: string; set: 'a Hash_set.t; logger: Logger.t}
+
+    let name {name; _} = name
+
+    let create (type elt) ~name ~logger
+        (module Elt : Hash_set.Elt_plain with type t = elt) : elt t =
+      let set =
+        Hash_set.create ~growth_allowed:true ?size:None (module Elt) ()
```

Comment:
it's a bit dangerous to allow this to have an unbounded size. We should have a maximum and a policy for handling what happens when it fills up (the policy maybe should be crash for now so we can catch bugs).